File: Solutions/115-distinct-subsequences/distinct-subsequences.ts
function numDistinct(s: string, t: string): number {
    // ------------------------------------------------------------
    // PROBLEM SUMMARY (IN COMMENTS ONLY, AS REQUESTED)
    // ------------------------------------------------------------
    // We are given two strings:
    //   - s : the source string
    //   - t : the target string
    //
    // We must count how many DISTINCT subsequences of s equal t.
    //
    // A subsequence:
    //   - Keeps relative order of characters
    //   - Can delete characters but NOT reorder them
    //
    // Example:
    //   s = "rabbbit"
    //   t = "rabbit"
    //
    // There are 3 different ways to delete characters from s
    // such that the remaining characters form "rabbit".
    //
    // ------------------------------------------------------------
    // DYNAMIC PROGRAMMING APPROACH
    // ------------------------------------------------------------
    //
    // Let dp[i][j] = number of ways to form
    //                t[0..j-1] using s[0..i-1]
    //
    // Meaning:
    //   i -> length considered from s
    //   j -> length considered from t
    //
    // Final answer will be dp[s.length][t.length]
    //
    // ------------------------------------------------------------
    // BASE CASES
    // ------------------------------------------------------------
    //
    // 1) dp[i][0] = 1 for all i
    //    - There is exactly ONE way to form an empty string ""
    //      from any prefix of s: by deleting everything.
    //
    // 2) dp[0][j] = 0 for all j > 0
    //    - We cannot form a non-empty t from an empty s.
    //
    // ------------------------------------------------------------
    // TRANSITION
    // ------------------------------------------------------------
    //
    // If s[i-1] === t[j-1]:
    //   We have TWO choices:
    //     1) Use this character -> dp[i-1][j-1]
    //     2) Skip this character -> dp[i-1][j]
    //
    //   dp[i][j] = dp[i-1][j-1] + dp[i-1][j]
    //
    // If s[i-1] !== t[j-1]:
    //   We can ONLY skip the character from s
    //
    //   dp[i][j] = dp[i-1][j]
    //
    // ------------------------------------------------------------
    // SPACE OPTIMIZATION
    // ------------------------------------------------------------
    //
    // Since dp[i] only depends on dp[i-1],
    // we can optimize from 2D -> 1D.
    //
    // dp[j] will represent dp for the "current row".
    //
    // IMPORTANT:
    //   We must iterate j from RIGHT TO LEFT
    //   to avoid overwriting values needed for dp[j-1].
    //
    // ------------------------------------------------------------

    const m = s.length; // length of source string
    const n = t.length; // length of target string

    // dp[j] = number of ways to form t[0..j-1]
    // from processed prefix of s
    const dp: number[] = new Array(n + 1).fill(0);

    // Base case:
    // There is exactly 1 way to form empty t ("")
    dp[0] = 1;

    // Loop through characters of s (outer loop)
    for (let i = 1; i <= m; i++) {

        // Loop through characters of t (inner loop)
        // RIGHT TO LEFT is CRITICAL
        for (let j = n; j >= 1; j--) {

            // If characters match, we can either:
            //   - use this character
            //   - or skip it
            if (s[i - 1] === t[j - 1]) {
                dp[j] = dp[j] + dp[j - 1];
            }

            // If characters do NOT match:
            //   dp[j] stays the same
            //   (we can only skip s[i - 1])
        }
    }

    // dp[n] now holds the number of distinct subsequences
    // of s that equal t
    return dp[n];
}

File: Solutions/115-distinct-subsequences/distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        
        # ---------------------------------------------------------
        # dp[j] = number of ways to form t[:j] using the characters
        #         of s that we have processed so far.
        #
        # For example:
        #
        # s = "babgbag"
        # t = "bag"
        #
        # dp[0] = 1 because there is exactly ONE way to form
        #         an empty string: choose nothing.
        #
        # dp[1] = number of ways to form "b"
        # dp[2] = number of ways to form "ba"
        # dp[3] = number of ways to form "bag"
        # ---------------------------------------------------------
        
        m = len(t)
        
        # We only need m + 1 positions because t has m characters.
        # Initially:
        #
        # dp = [1, 0, 0, ..., 0]
        #
        # dp[0] = 1
        # dp[j] = 0 for j > 0
        #
        # This means:
        # - There is 1 way to create ""
        # - There are 0 ways to create a non-empty string
        #   before processing any characters from s.
        dp = [0] * (m + 1)
        dp[0] = 1
        
        # ---------------------------------------------------------
        # Process every character of s one at a time.
        # ---------------------------------------------------------
        for ch in s:
            
            # -----------------------------------------------------
            # We iterate BACKWARDS through t.
            #
            # Why backwards?
            #
            # Because dp[j - 1] represents the number of ways to
            # form t[:j-1] BEFORE using the current character of s.
            #
            # If we went forwards, dp[j - 1] might already have
            # been updated using the current character, which would
            # accidentally allow us to use the same character from
            # s more than once.
            #
            # Example:
            #
            # s = "aa"
            # t = "aa"
            #
            # When processing the second 'a', we want to use the
            # FIRST 'a' to form "a", and the SECOND 'a' to complete
            # "aa".
            #
            # Going backwards guarantees this.
            # -----------------------------------------------------
            for j in range(m, 0, -1):
                
                # -------------------------------------------------
                # If the current character from s matches the
                # character we need in t, we have a choice:
                #
                # 1. USE this character from s
                #
                #    dp[j - 1] ways of forming t[:j-1]
                #    become ways of forming t[:j].
                #
                # 2. DON'T USE this character
                #
                #    dp[j] already contains all the ways we had
                #    before seeing this character.
                #
                # So if characters match:
                #
                # dp[j] += dp[j - 1]
                #
                # -------------------------------------------------
                if ch == t[j - 1]:
                    dp[j] += dp[j - 1]
        
        # ---------------------------------------------------------
        # dp[m] contains the number of ways to form the ENTIRE
        # string t using characters from s.
        # ---------------------------------------------------------
        return dp[m]
